docs/dev/commitlog-file-format.md

```diff
@@ -62,4 +62,55 @@ Version 2
         size            : size of all entries in this multi-entry + headers
         crc             : CRC32 of magic, size
         <entries> * N
+        crc2            : CRC32 of magic, size and data in each entry
+```
+
+Version 3
+---------
+
+Modified from v3 to improve error detection/false positive elimination.
+This version does CRC per written disk block instead of actual entry.
+Every block is also tagged with which file is being written, this to
+be able to better distinguish new data from that left over from recycling
+(reusing old files) and actual disk/file corruption.
+
+```
+
+        Segment file header
+
+        magic           : uint32_t - ('S'<<24) |('C'<< 16) | ('L' << 8) | 'C';
+        version         : uint32_t - same as descriptor
+        id              : uint64_t - same as descriptor
+        alignment       : uint32_t - disk block size
+        crc             : uint32_t - CRC32 of version, low 32 of id, high 32 of id. 
+
+        Chunk header
+
+        file_pos        : uint32_t - the file position of next chunk
+        crc             : uint32_t - CRC32 of low 32 of segment id, high 32 of id and file offset of end of this header.
+
+        Entry
+
+        size            : uint32_t - size of entry (data + full headers). Must be smaller than MAX_UINT32.
+        crc             : uint32_t - CRC32 of size
+        data            : bytes - actual entry data
+
+        Multi-entry
+
+        magic           : multi marker - 0xffffffff (MAX_UINT32)
+        size            : size of all entries in this multi-entry + headers
+        crc             : CRC32 of magic, size
+        <entries> * N
+
+        Disk block (block size = `alignment`)
+
+        0 - <bs - 12>   : Interleaved file data, i.e. the content above
+        <bs - 12>       : uint32_t - CRC32 of data, low 32 of segment id, high 32 of segment id
```

Comment:
Each block should also have the generation, no? So we can distinguish between a correctly-written block of the previous generation (which signals end-of-stream) and a correctly-written block of this generation.

Reply:
Segment ID == gen.

Reply:
yes of course.